lib/spack/env/cc

```diff
@@ -114,104 +114,118 @@ case "$command" in
 esac
 
 # Finish setting up the mode.
-if [ -z "$mode" ]; then
+if [[ -z $mode ]]; then
     mode=ccld
     for arg in "$@"; do
-        if [ "$arg" = -v -o "$arg" = -V -o "$arg" = --version -o "$arg" = -dumpversion ]; then
-            mode=vcheck
-            break
-        elif [ "$arg" = -E ]; then
-            mode=cpp
-            break
-        elif [ "$arg" = -c ]; then
-            mode=cc
-            break
-        fi
+        case "$arg" in
+            -v|-V|--version|-dumpversion)
+                mode=vcheck
+                break
+                ;;
+            -E)
+                mode=cpp
+                break
+                ;;
+            -c)
+                mode=cc
+                break
+                ;;
+            -S)
+                mode=as
+                break
+                ;;
+        esac
     done
 fi
 
 # Dump the version and exist if we're in testing mode.
-if [ "$SPACK_TEST_COMMAND" = "dump-mode" ]; then
+if [[ $SPACK_TEST_COMMAND = dump-mode ]]; then
     echo "$mode"
     exit
 fi
 
 # Check that at least one of the real commands was actually selected,
 # otherwise we don't know what to execute.
-if [ -z "$command" ]; then
+if [[ -z $command ]]; then
     die "ERROR: Compiler '$SPACK_COMPILER_SPEC' does not support compiling $language programs."
 fi
 
 # Save original command for debug logging
 input_command="$@"
+args=("$@")
 
-#
-# Now do real parsing of the command line args, trying hard to keep
-# non-rpath linker arguments in the proper order w.r.t. other command
-# line arguments.  This is important for things like groups.
-#
-includes=()
-libraries=()
-libs=()
-rpaths=()
-other_args=()
+# Dump parsed values for unit testing if asked for
+if [[ -n $SPACK_TEST_COMMAND ]]; then
```

Comment:
The problem with this if statement is that it makes the debug/test output useless, because it doesn't reflect what the parsing is actually doing when a real build takes place.  The script basically operates in an entirely different way when `SPACK_TEST_COMMAND` is set and when it isn't.  

The thing to do here is not to duplicate the parsing, but to ensure that `test/cc.py` tests the important aspects of the parsing.  If it's a pain to extract the different types of args independently, you could reduce the test commands to "dump args" and have `test/cc.py` ensure that the includes, libs, linker args, etc. are all a) present and b) appear in the correct order.  That would probably be a better test than what's there now, since obviously it currently messes up the ordering (which is important!).

If the test were updated (and the `cc` script simplified as a result) then I think this would be easy to merge.  We have another internal patch that conflicts with this that I have also been putting off merging.  This is why it took me so long (sorry!).  That patch adds better handling of `.a` files (which I believe are also relevant for groups).  I can add that after this is merged if the testing issue gets fixed...